### ISSUE TITLE
contender daemon (contenderd)

### ISSUE THREAD

**Issue by zeroXbrock** [flashbots/contender#58]
At the time of writing, contender is really good at running a single, well-defined scenario. We can say something like "deploy uniswapV2, create some tokens and make trading pairs for them, then spam the network with swaps at 100 swaps per block (or per second)" and contender will do it quite well. But it has some drawbacks:
1. We can't variate the frequency of transaction requests -- for the whole duration of the spam run, you're sending at a fixed rate. This isn't exactly realistic. In prod, the systems we're testing will experience lulls and bursts of activity; that's what we really need to test.
2. Scenario definitions (TOML config files) don't scale well. They can already get messy when we try to make the scenario more complex (look at the [UniV2 scenario](https://github.com/flashbots/contender/blob/main/scenarios/univ2ConfigTest.toml) for an idea of what I mean). If we try to pack new functions and features into the existing system, we might lose some of the system's simplicity.
3. The CLI doesn't scale well. We can do a single spam run; fixed-duration, narrowly-scoped; but then we have to wait for that run to complete to get the results. A single failed tx can crash the run (an outstanding issue in contender-core which I'll fix soon), so we need to start by dialing in the maximum output of contender to match what the node can handle, and then manually spam with those parameters over and over.

I have an idea that I think can lay the foundation to fix all these problems: a **web API ("`contenderd`")** (maybe running JSON-RPC) that runs features from contender-core. We'd essentially rewrite the CLI as a web API (with some obvious differences). Using this API, it would be easy to build new, more advanced executors that run existing scenarios (short-duration, narrowly-scoped spam definitions) in larger numbers and in parallel, yielding a more-realistic ensemble of spam, while also simplifying the task of defining complex scenarios (just use several config files).

I haven't specced out the API yet but off the top of my head, I see a few key differences in how it would work:
- TOML config files might need to be translated to JSON (easy; serde has us covered)
- it's inefficient and less-than-ergonomic to send the whole scenario config over and over. It would be more comfy to register scenarios individually via the API, and then have the "spam API" methods run registered scenarios
- private keys are needed no matter what, but it would be nice to simply register your private key with the system, and then use that as the default, so you don't have to keep sending your key over the wire

**Comment by ferranbt** [comment]
> The CLI doesn't scale well. We can do a single spam run; fixed-duration, narrowly-scoped; but then we have to wait for that run to complete to get the results. A single failed tx can crash the run (an outstanding issue in contender-core which I'll fix soon), so we need to start by dialing in the maximum output of contender to match what the node can handle, and then manually spam with those parameters over and over.

Agree on this one.

> We can't variate the frequency of transaction requests -- for the whole duration of the spam run, you're sending at a fixed rate. This isn't exactly realistic. In prod, the systems we're testing will experience lulls and bursts of activity; that's what we really need to test.

How do you imagine the system, you have the cli to update in real time the frequency of the daemon? How would this affect the reproducibility of the experiment?

> Scenario definitions (TOML config files) don't scale well. They can already get messy when we try to make the scenario more complex (look at the [UniV2 scenario](https://github.com/flashbots/contender/blob/main/scenarios/univ2ConfigTest.toml) for an idea of what I mean). If we try to pack new functions and features into the existing system, we might lose some of the system's simplicity.

Do you get rid of this? I would imagine you still need to do this somewhere.

> - it's inefficient and less-than-ergonomic to send the whole scenario config over and over. It would be more comfy to register scenarios individually via the API, and then have the "spam API" methods run registered scenarios
> - private keys are needed no matter what, but it would be nice to simply register your private key with the system, and then use that as the default, so you don't have to keep sending your key over the wire

What would be the new DevEx of this stateful system? Would you have to setup everything on the daemon and then run the experiment?

Who would be the client of this API? Would that still be contender so that the current calls (i.e. spam) are now a call to an API?

It would be nice to have an example of the new API calls.

**Comment by zeroXbrock** [comment]
@ferranbt 

I spent about 3 hours answering your questions and deliberating, and now I'm thinking a web API is just not a good idea. This would never be hosted online, so why bother?

I was thinking that a web API would make it easier and more accessible to build new features, but who would be building them? Might as well just build the features into the CLI.

Scrapped.
